synapse/lib/blobfile.py

```diff
@@ -0,0 +1,310 @@
+            None
+        '''
+        self._writeoff(off, byts)
+
+    def alloc(self, size):
+        '''
+        Allocate a block within the Blob and return the offset.
+
+        Args:
+            size (int): Number of bytes to allocate within the blob.
+
+        Example:
+            Store a string by allocating space for it in a blob file:
+
+                s = 'foobar'
+                byts = s.encode()
+                off = blob.alloc(len(byts))
+                blob.writeoff(off, byts)
+
+        Returns:
+            int: Offset within the blob to use to store size bytes.
+        '''
+        # Account for the blob header
+        fullsize = headsize + size
+
+        with self.alloclock:
+
+            # Fire our alloc event
+            self.fire('blob:alloc', size=size)
+
+            nsize = self._size + fullsize
+
+            # Grow the file
+            self.atom.resize(nsize)
+
+            # Write the size dword
+            self.writeoff(self._size, struct.pack(headfmt, size))
+
+            # Compute the return value
+            dataoff = self._size + headsize
+
+            # Update our runtime size
+            self._size = nsize
+
+        return dataoff
+
+    def size(self):
+        '''
+        Get the amount of space currently used by the blob.
+
+        Returns:
+            int: Size of the bytes currently allocated by the blob.
+        '''
+        return self._size
+
+    def walk(self, callback=None):
+        '''
+        Walk the blobfile from the first record until the end.
+
+        Args:
+            callback (func): A callback function which is executed when a
+            BlobFile record is encountered.  The callback should accept the
+            following function signature: ``(fd, baseoff, off, size)``.  fd is
+            the blob's file descriptor; baseoff is the offset where the blob
+            header is located, offset is the location of the blob data, and
+            size is the number of bytes allocated for the blob. If no callback
+            is provided, nothing will happen when a BlobFile record is
+            encountered.
+
+        Notes:
+            In addition to the callback, when walking the blobfile additional
+            conditions may be encountered. These will fire an event and exit
+            the walk() operation.
+
+            These events will not serialize since they contain a
+            reference to the underlying file descriptor and are designed in
+            order to allow creation of tools which can inspect/consume bytes
+            from a BlobFile.
+
+                blob:walk:unpkerr - Fired if there is an error unpackting the Blobfile header.
+
+                blob:walk:truncated - Fired if there is a truncated file encountered.
+
+                blob:walk:done - Fired when the BlobWalker is done walking the file.
+
+        Returns:
+            None
+        '''
+        # Snag the max filesize
+        self.fd.seek(0, os.SEEK_END)
```

Comment:
Almost nothing should use self.fd as the internal offset is not guaranteed atomic and could cause some *nasty* multi-threading issues.  By using self.atom, the seek/tell/read calls here can all become readoff() calls which may or may not use a lock, depending on the optimal platform AtomFile ( hopefully pread ).

Reply:
I'd consider renaming `self.fd` to `self._fd` and adding a comment where it is initialized